### PR TITLE
Rely on PGO for isinst/castclass

### PR DESCRIPTION
**Pull request by EgorBo**
Follow up to https://github.com/dotnet/runtime/pull/65460 (and [d-o](https://dev.azure.com/dnceng/internal/_git/dotnet-optimization/pullrequest/21334?_a=files))

Example (with `DOTNET_JitCastProfiling=1`):
```csharp
using System.Runtime.CompilerServices;
using System.Threading;

public interface IClass {}
public class ClassA : IClass {}
public class ClassB : ClassA {}


public class Program
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static ClassA CastToClassA(object o) => (ClassA)o; // currently we always emit a fast path for ClassA here
                                                       // but in my case o is always ClassB (PGO helps)

    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool IsClassA(object o) => o is ClassA;     // we don't expand it without PGO

    [MethodImpl(MethodImplOptions.NoInlining)]
    static IClass CastToIClass(object o) => (IClass)o; // we don't expand it without PGO

    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool IsIClass(object o) => o is IClass;     // we don't expand it without PGO


    public static void Main()
    {
        // promote methods to tier1
        var b = new ClassB();
        for (int i = 0; i < 100; i++)
        {
            CastToClassA(b);
            IsClassA(b);
            CastToIClass(b);
            IsIClass(b);
            Thread.Sleep(16);
        }
    }
}
```
Codegen diff: https://www.diffchecker.com/RFblv9RB

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -2141,21 +2141,17 @@ bool Compiler::impIsCastHelperEligibleForClassProbe(GenTree* tree)
 // Returns:
 //    true if the tree is a cast helper with potential profile data
 //
-bool Compiler::impIsCastHelperMayHaveProfileData(GenTree* tree)
+bool Compiler::impIsCastHelperMayHaveProfileData(CorInfoHelpFunc helper)
 {
-    if (!opts.jitFlags->IsSet(JitFlags::JIT_FLAG_BBOPT) || (JitConfig.JitCastProfiling() != 1))
```
Comment (Wraith2) [inline src/coreclr/jit/importer.cpp:2146]: Why is this no longer needed?

Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:2146]: However, the built-in profile StandardOptimization.mibc is expected to have probes for isisnst/castclass

**Comment by EgorBo** [comment]
@AndyAyersMS @jakobbotsch PTAL

This PR consumes profile data from mibc for casts/isinst. I guarded it with `DOTNET_JitConsumeProfileForCasts` just to wait till dotnet-optimization properly propagates mibc with casts/isinst.

**Comment by EgorBo** [comment]
@AndyAyersMS @jakobbotsch PTAL

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -11636,6 +11640,61 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
...
+                                                    fgPgoData, ilOffset);
+            }
+
+            if (likelyClassCount > 0)
+            {
+                LikelyClassRecord    likelyClass = likelyClasses[0];
+                CORINFO_CLASS_HANDLE likelyCls   = likelyClass.clsHandle;
+
+                if ((likelyCls != NO_CLASS_HANDLE) &&
+                    (likelyClass.likelihood > (UINT32)JitConfig.JitGuardedDevirtualizationChainLikelihood()))
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:11681]: Should this rather get its own config variable?

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:11681]: Just didn't want to produce even more variables 😄 and the default value for this one was OK to me.
in my other PR to add "multiple guesses" I slightly changed the whole logic so I'll leave it as is for now

**Pull request merged**